pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.java

```diff
@@ -279,55 +278,68 @@ public ApexTreeBuilder(String sourceCode) {
     private void buildFormalComment(AstNode node) {
         if (parents.peek() == node) {
             ApexNode<?> parent = (ApexNode<?>) nodes.peek();
-            TokenLocation tokenLocation = getApexDocTokenLocation(getApexDocIndex(parent));
+            ApexDocTokenLocation tokenLocation = getApexDocTokenLocation(getApexDocIndex(parent));
             if (tokenLocation != null) {
                 ASTFormalComment comment = new ASTFormalComment(tokenLocation.token);
                 comment.calculateLineNumbers(sourceCodePositioner, tokenLocation.index,
                         tokenLocation.index + tokenLocation.token.length());
+
+                // move existing nodes so that we can insert the comment as the first node
+                for (int i = parent.jjtGetNumChildren(); i > 0; i--) {
+                    parent.jjtAddChild(parent.jjtGetChild(i - 1), i);
+                }
+
                 parent.jjtAddChild(comment, 0);
                 comment.jjtSetParent(parent);
             }
         }
     }
 
     private int getApexDocIndex(ApexNode<?> node) {
-        final int index = node.getNode().getLoc().getStartIndex();
+        ASTAnnotation annotation = node.getFirstDescendantOfType(ASTAnnotation.class);
+        ApexNode<?> firstNode = annotation == null ? node : annotation;
+        final int index = firstNode.getNode().getLoc().getStartIndex();
         return sourceCode.lastIndexOf('\n', index);
     }
 
-    private TokenLocation getApexDocTokenLocation(int index) {
```

Comment:
There seems to be an easier solution, which avoids dealing with Annotation nodes specifically (maybe other nodes have also a different position in the AST?).

So, yes, we can't use ListIterator, since we might skip more than one token until we find a match and then we don't backup more than one... so in the end we loose a comment.

But we can iterate through the comments and find the nearest match. In order to "mark" the comments, that are already matched, we remove them from the list. Then for the next call, we start the search from the beginning again. With the list iterator, we started the next search after the last match, which is a problem, if the order of the AST doesn't match the source code.

```java
    private TokenLocation getApexDocTokenLocation(int index) {
        TokenLocation last = null;
        for (TokenLocation location : apexDocTokenLocations) {
            if (location.index >= index) {
                // gone too far - this token corresponds to a different node
                break;
            }
            last = location;
        }
        if (last != null) {
            apexDocTokenLocations.remove(last);
        }
        return last;
    }
```

If the list is big, I think, it still takes some time to copy the content when removing a item in the ArrayList. Since we mostly remove a item from the beginning, it should be found fast. We probably should use a `LinkedList` in `buildApexDocTokenLocations()`.